## Hard Drive Failure Analytics Analytics - Exploratory Data Analysis
Purpose: The purpose of this notebook is to review the source data and create step by step process and extracting the data from online source and reviewing the data locally

Objectives:
- Extract zipfiles from source
- Uzip files and import CSV files
- Describe data (columns, schema, datatypes, total row count, column count etc.)

In [1]:
import click
from glob import glob
import os
import pandas as pd
import pyarrow.parquet as pq
import requests
from sqlalchemy import create_engine
import zipfile
import tempfile    # For creating sandbox directories
import shutil      # For moving files safely

/home/user1129/MyDocuments/10_14_Life_Admin/13_Technology/100_Coding/100_01_Python/data_engineering/hard_drive_failure_analytics_dashboard/.venv/lib/python3.13/site-packages/requests/__init__.py:113: RequestsDependencyWarning: urllib3 (2.6.3) or chardet (6.0.0.post1)/charset_normalizer (3.4.6) doesn't match a supported version!
  warnings.warn(


Backblaze datasets are organized as the following:
- 2013/2014 - full annual zip file with multiple csv files. Url suffix ends as **data_{year}.zip**
- 2015 and onward - separated as quarterly zip files with multiple csv files (4 quarters for each year). Url suffix ends as **data_{quarter}_{year}.zip**

For testing let's pull just 2025 and see what we have.

Minigoals:
- Extract one zip file at time -> reduce number of data fields -> read csv files as one

In [7]:
year = 2025
qtr = "1"
file_suffix = f"{"data_Q"+qtr}_{year}"

In [3]:
# Creating helpful functions

def download_and_extract(zip_url, extract_dir='../data/zip'):    
    # 1. Ensure ./data/zip folder exists
    os.makedirs(extract_dir, exist_ok=True)

    # Get filename from URL path
    zip_filename = zip_url.split('/')[-1] or 'archive.zip'

    # 2. Download zip file from URL and save to extract_dir
    zip_path = os.path.join(extract_dir, zip_filename)
    
    print(f"Downloading {zip_url} to {zip_path}...")
    response = requests.get(zip_url, stream=True)
    response.raise_for_status()  # Raise HTTPError for bad responses
    
    with open(zip_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            if chunk:
                f.write(chunk)
    
    return zip_path

def unzip_file(input_file, output_dir=f"../data/raw/{year}"):
    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)
    
    # Validate input file exists
    if not os.path.isfile(input_file):
        raise FileNotFoundError(f"Input file does not exist: {input_file}")
    
    with zipfile.ZipFile(input_file, 'r') as zip_ref:
        # List contents (optional check)
        print("Contents:", zip_ref.namelist())
        
        # Extract all files to the specified output directory
        zip_ref.extractall(output_dir)
        print(f"Extracted to: {output_dir}")


In [ ]:
# dev test link
base_url = "https://github.com/ammartin8/hard_drive_analytics_dashboard/releases/download/hd_2025_Q1"

In [29]:
def safe_extract(zip_path, extract_to):
    if not os.path.exists(zip_path):
        raise FileNotFoundError(f"ZIP file not found: {zip_path}")
    
    # 1. Create a temporary sandbox directory
    # This prevents extracting malicious files outside intended folder
    temp_dir = tempfile.mkdtemp(prefix='safe_extract_')
    print(f"[+] Created temp sandbox: {temp_dir}")
    
    try:
        with zipfile.ZipFile(zip_path, 'r') as z:
            # Extract to the temporary safe dir first
            if not z.namelist():
                raise ValueError("ZIP file is empty or invalid.")
            
            # Extract all contents to temporary directory
            print(f"[i] Extracting to sandbox...")
            z.extractall(path=temp_dir)
            
            # 2. Move valid contents to target ONLY if needed (if using a zip as a package)
            # This step usually depends on your deployment logic.
            for root, dirs, files in os.walk(temp_dir):
                relative_path = os.path.relpath(root, temp_dir)
                
                # Ensure relative path is not empty and doesn't contain '..' 
                if relative_path == '.' or '..' not in relative_path:
                    dest_root = os.path.join(extract_to, relative_path)
                    os.makedirs(dest_root, exist_ok=True)
                    for file in files:
                        src_file = os.path.join(root, file)
                        try:
                            shutil.move(src_file, dest_root + "/" + file)
                            print(f'File: {file} moved to {dest_root}')
                        except Exception as e:
                            print(f"Failed to move {file}: {e}")
    except Exception as e:
        print(f"An error occurred: {e}")

        

In [22]:
# Extract data
# download_and_extract(base_url+f"{file_suffix}.zip")

In [23]:
# unzip_file(f"../data/zip/{file_suffix}.zip")

In [ ]:
# safe_extract('../data/zip/data_Q1_2025.zip', CSV_DIR)

[+] Created temp sandbox: /tmp/safe_extract_innqyf4p
[i] Extracting to sandbox...
File: 2025-02-20.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-15.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-31.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-03-28.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-02-06.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-03-06.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-02-19.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-11.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-03-27.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-03-07.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-13.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-02-12.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-14.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-01-27.csv moved to ../data/source_csv/data_Q1_2025
File: 2025-02-13.csv moved to ../data/source_csv/d

In [9]:
# Read quick sample
df_test = pd.read_csv(f"../data/source_csv/{file_suffix}/{year}-01-01.csv"
    , nrows=0)
df_test.head()

,date,serial_number,model,capacity_bytes,failure,datacenter,cluster_id,vault_id,pod_id,pod_slot_num,...,smart_250_normalized,smart_250_raw,smart_251_normalized,smart_251_raw,smart_252_normalized,smart_252_raw,smart_254_normalized,smart_254_raw,smart_255_normalized,smart_255_raw


In [6]:
os.path.basename("./data/zip/{file_suffix}.zip")

'{file_suffix}.zip'

In [10]:
files = [f for f in os.listdir("../data/source_csv/2025/data_Q1_2025/")]
print(files)

['2025-02-20.csv', '2025-01-15.csv', '2025-01-31.csv', '2025-03-28.csv', '2025-02-06.csv', '2025-03-06.csv', '2025-02-19.csv', '2025-01-11.csv', '2025-03-27.csv', '2025-03-07.csv', '2025-01-13.csv', '2025-02-12.csv', '2025-01-14.csv', '2025-01-27.csv', '2025-02-13.csv', '2025-03-14.csv', '2025-03-01.csv', '2025-01-16.csv', '2025-01-12.csv', '2025-01-30.csv', '2025-01-07.csv', '2025-03-31.csv', '2025-01-05.csv', '2025-03-20.csv', '2025-03-15.csv', '2025-02-25.csv', '2025-01-29.csv', '2025-01-26.csv', '2025-01-03.csv', '2025-01-02.csv', '2025-01-24.csv', '2025-03-12.csv', '2025-02-17.csv', '2025-02-28.csv', '2025-03-10.csv', '2025-03-24.csv', '2025-03-17.csv', '2025-02-18.csv', '2025-01-21.csv', '2025-02-04.csv', '2025-03-03.csv', '2025-01-20.csv', '2025-01-19.csv', '2025-01-09.csv', '2025-02-14.csv', '2025-02-07.csv', '2025-03-16.csv', '2025-03-02.csv', '2025-03-13.csv', '2025-03-29.csv', '2025-01-25.csv', '2025-01-01.csv', '2025-01-17.csv', '2025-02-24.csv', '2025-01-06.csv', '2025-01-

In [31]:
DATA_ROOT = "../data"
CSV_DIR = os.path.join(DATA_ROOT, "source_csv")

In [25]:
df_test.dtypes

date                    object
serial_number           object
model                   object
capacity_bytes          object
failure                 object
                         ...  
smart_252_raw           object
smart_254_normalized    object
smart_254_raw           object
smart_255_normalized    object
smart_255_raw           object
Length: 197, dtype: object

In [26]:
df_test.shape

(0, 197)

## Notes
- The csv file sizes are too large to combine as one df for an entire quarter of data. Load into postgres in chunks for analysis instead.
- Since we are focused on failure rates, there are specific SMART attributes that were identifed as key to indicating failure or soon to be failure drives (refernced from backblaze):
    - SMART 5: Reallocated_Sector_Count.
	- SMART 187: Reported_Uncorrectable_Errors.
	- SMART 188: Command_Timeout.
	- SMART 197: Current_Pending_Sector_Count.
	- SMART 198: Offline_Uncorrectable.

In [11]:
dtype = {
    "serial_number": "string",
    "model": "string",
    "capacity_bytes": "Int64",
    "failure": "Int64",
    "datacenter": "string",
    "cluster_id": "Int64",
    "vault_id": "Int64",
    "pod_id": "Int64",
    "pod_slot_num": "float64",
    "is_legacy_format": "boolean",
    "smart_5_normalized": "float64",  # SMART attribute (normalized value)
    "smart_5_raw": "float64", # Raw SMART value
    "smart_187_normalized": "float64",
    "smart_187_raw": "float64",
    "smart_188_normalized": "float64",
    "smart_188_raw": "float64",
    "smart_197_normalized": "float64",
    "smart_197_raw": "float64",
    "smart_198_normalized": "float64",
    "smart_198_raw": "float64"
}

parse_dates = [
    "date"
]

In [13]:
col_to_keep = []
for col in df_test.columns:
    if col in dtype or col in parse_dates:
        col_to_keep.append(col)
print(col_to_keep)

['date', 'serial_number', 'model', 'capacity_bytes', 'failure', 'datacenter', 'cluster_id', 'vault_id', 'pod_id', 'pod_slot_num', 'is_legacy_format', 'smart_5_normalized', 'smart_5_raw', 'smart_187_normalized', 'smart_187_raw', 'smart_188_normalized', 'smart_188_raw', 'smart_197_normalized', 'smart_197_raw', 'smart_198_normalized', 'smart_198_raw']


In [14]:
for col in dtype:
    print(col)

serial_number
model
capacity_bytes
failure
datacenter
cluster_id
vault_id
pod_id
pod_slot_num
is_legacy_format
smart_5_normalized
smart_5_raw
smart_187_normalized
smart_187_raw
smart_188_normalized
smart_188_raw
smart_197_normalized
smart_197_raw
smart_198_normalized
smart_198_raw


In [22]:
df = pd.read_csv(
    f"../data/source_csv/{file_suffix}/{year}-01-01.csv"
    , nrows=1
    , dtype=dtype
    ,parse_dates=parse_dates
    ,usecols=col_to_keep
).head(0)

In [24]:
df

,date,serial_number,model,capacity_bytes,failure,datacenter,cluster_id,vault_id,pod_id,pod_slot_num,...,smart_5_normalized,smart_5_raw,smart_187_normalized,smart_187_raw,smart_188_normalized,smart_188_raw,smart_197_normalized,smart_197_raw,smart_198_normalized,smart_198_raw


In [23]:
df.dtypes

date                    datetime64[us]
serial_number                   string
model                           string
capacity_bytes                   Int64
failure                          Int64
datacenter                      string
cluster_id                       Int64
vault_id                         Int64
pod_id                           Int64
pod_slot_num                   float64
is_legacy_format               boolean
smart_5_normalized             float64
smart_5_raw                    float64
smart_187_normalized           float64
smart_187_raw                  float64
smart_188_normalized           float64
smart_188_raw                  float64
smart_197_normalized           float64
smart_197_raw                  float64
smart_198_normalized           float64
smart_198_raw                  float64
dtype: object

### Setting up postgres database

In [15]:
# Create sql engine
engine = create_engine('postgresql+psycopg://root:root@localhost:5432/hard_drive_db')

In [16]:
print(pd.io.sql.get_schema(df, name='hard_drive_data', con=engine))


CREATE TABLE hard_drive_data (
	date TIMESTAMP WITHOUT TIME ZONE, 
	serial_number TEXT, 
	model TEXT, 
	capacity_bytes BIGINT, 
	failure BIGINT, 
	datacenter TEXT, 
	cluster_id BIGINT, 
	vault_id BIGINT, 
	pod_id BIGINT, 
	pod_slot_num FLOAT(53), 
	is_legacy_format BOOLEAN, 
	smart_5_normalized FLOAT(53), 
	smart_5_raw FLOAT(53), 
	smart_187_normalized FLOAT(53), 
	smart_187_raw FLOAT(53), 
	smart_188_normalized FLOAT(53), 
	smart_188_raw FLOAT(53), 
	smart_197_normalized FLOAT(53), 
	smart_197_raw FLOAT(53), 
	smart_198_normalized FLOAT(53), 
	smart_198_raw FLOAT(53)
)




In [17]:
# Create empty table with data fields
first = True
if first:
    df.head(n=0).to_sql(name='hard_drive_data', con=engine, if_exists='replace')
    print("Table created")
    first = False

0

In [5]:
files = sorted(os.listdir(f"../data/source_csv/data_Q1_2025"))
print(files)

['2025-01-01.csv', '2025-01-02.csv', '2025-01-03.csv', '2025-01-04.csv', '2025-01-05.csv', '2025-01-06.csv', '2025-01-07.csv', '2025-01-08.csv', '2025-01-09.csv', '2025-01-10.csv', '2025-01-11.csv', '2025-01-12.csv', '2025-01-13.csv', '2025-01-14.csv', '2025-01-15.csv', '2025-01-16.csv', '2025-01-17.csv', '2025-01-18.csv', '2025-01-19.csv', '2025-01-20.csv', '2025-01-21.csv', '2025-01-22.csv', '2025-01-23.csv', '2025-01-24.csv', '2025-01-25.csv', '2025-01-26.csv', '2025-01-27.csv', '2025-01-28.csv', '2025-01-29.csv', '2025-01-30.csv', '2025-01-31.csv', '2025-02-01.csv', '2025-02-02.csv', '2025-02-03.csv', '2025-02-04.csv', '2025-02-05.csv', '2025-02-06.csv', '2025-02-07.csv', '2025-02-08.csv', '2025-02-09.csv', '2025-02-10.csv', '2025-02-11.csv', '2025-02-12.csv', '2025-02-13.csv', '2025-02-14.csv', '2025-02-15.csv', '2025-02-16.csv', '2025-02-17.csv', '2025-02-18.csv', '2025-02-19.csv', '2025-02-20.csv', '2025-02-21.csv', '2025-02-22.csv', '2025-02-23.csv', '2025-02-24.csv', '2025-02-

In [ ]:
# Ingest data in file chunks
for data_file in files:
    df = pd.read_csv(
        f"../data/source_csv/{file_suffix}/{data_file}",
        dtype=dtype,
        parse_dates=parse_dates,
        usecols = col_to_keep
    )
    df.to_sql(name='hard_drive_data', con=engine, if_exists='append')
    print(f"Loaded {len(df)} records from file: {data_file}")

Loaded 304957 records from file: 2025-01-01.csv
Loaded 304842 records from file: 2025-01-02.csv
Loaded 304848 records from file: 2025-01-03.csv
Loaded 304922 records from file: 2025-01-04.csv
Loaded 304897 records from file: 2025-01-05.csv
Loaded 305135 records from file: 2025-01-06.csv
Loaded 305211 records from file: 2025-01-07.csv
Loaded 305243 records from file: 2025-01-08.csv
Loaded 305248 records from file: 2025-01-09.csv
Loaded 305255 records from file: 2025-01-10.csv
Loaded 305156 records from file: 2025-01-11.csv
Loaded 305017 records from file: 2025-01-12.csv
Loaded 305072 records from file: 2025-01-13.csv
Loaded 305230 records from file: 2025-01-14.csv
Loaded 305105 records from file: 2025-01-15.csv
Loaded 305074 records from file: 2025-01-16.csv
Loaded 305186 records from file: 2025-01-17.csv
Loaded 305141 records from file: 2025-01-18.csv
Loaded 305105 records from file: 2025-01-19.csv
Loaded 305122 records from file: 2025-01-20.csv
Loaded 305113 records from file: 2025-01

### Creating One Ingestion Script

In [47]:
# refactor code to accept year and quarter parameters
def download_data(year=2025, qtr=1):    
    # Ensure ../data/zip folder exists
    download_dir='../data/zip'
    os.makedirs(download_dir, exist_ok=True)

    # Define base url
    base_url = "https://f001.backblazeb2.com/file/Backblaze-Hard-Drive-Data"

    # File structure
    file_suffix = "data_Q"+f"{qtr}_{year}"

    zip_url = f"{base_url}/{file_suffix}.zip"
    # Get filename from URL path
    zip_filename = zip_url.split('/')[-1] or 'archive.zip'

    # Download zip file from URL and save to download_dir
    zip_path = os.path.join(download_dir, zip_filename)
    
    print(f"Downloading {zip_url} to {zip_path}...")
    response = requests.get(zip_url, stream=True)
    response.raise_for_status()  # Raise HTTPError for bad responses
    
    with open(zip_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            if chunk:
                f.write(chunk)
    print(f"{zip_path} is downloaded!")
    return zip_path

In [ ]:
# download_data_path = download_data(year=2025, qtr=1)
print(download_data_path)

../data/zip/data_Q1_2025.zip is downloaded!
../data/zip/data_Q1_2025.zip


In [19]:
# Data must be structured as data_{qtr}_{year}.zip
print(download_data_path.split('/')[-1].split('.')[0])

NameError: name 'download_data_path' is not defined

In [ ]:
# refactor code to accept previously generated download file path
def unzip_file(input_file):
    
    # Extract year and qtr from zip_path to create folder names
    # Note: file name must be structured as ../data/zip/data_{qtr}_{year}.zip for year/qtr to work
    zip_path_year = download_data_path.split('/')[-1].split('_')[-1].split('.')[0]

    # Ensure extract directory exists
    extract_dir=f"../data/source_csv/{zip_path_year}"
    os.makedirs(extract_dir, exist_ok=True)
    
    # Validate input file exists
    if not os.path.isfile(input_file):
        raise FileNotFoundError(f"Input file does not exist: {input_file}")
    
    with zipfile.ZipFile(input_file, 'r') as zip_ref:
        # List contents (optional check)
        print("Contents:", zip_ref.namelist())
        
        # Extract all files to the specified extract directory
        zip_ref.extractall(extract_dir)
        print(f"A total of {len(zip_ref.namelist())} file(s) extracted.") 
        print(f"Extracted to: {extract_dir}")

In [ ]:
# unzip_file()

Contents: ['data_Q1_2025/2025-01-01.csv', 'data_Q1_2025/2025-01-02.csv', 'data_Q1_2025/2025-01-03.csv', 'data_Q1_2025/2025-01-04.csv', 'data_Q1_2025/2025-01-05.csv', 'data_Q1_2025/2025-01-06.csv', 'data_Q1_2025/2025-01-07.csv', 'data_Q1_2025/2025-01-08.csv', 'data_Q1_2025/2025-01-09.csv', 'data_Q1_2025/2025-01-10.csv', 'data_Q1_2025/2025-01-11.csv', 'data_Q1_2025/2025-01-12.csv', 'data_Q1_2025/2025-01-13.csv', 'data_Q1_2025/2025-01-14.csv', 'data_Q1_2025/2025-01-15.csv', 'data_Q1_2025/2025-01-16.csv', 'data_Q1_2025/2025-01-17.csv', 'data_Q1_2025/2025-01-18.csv', 'data_Q1_2025/2025-01-19.csv', 'data_Q1_2025/2025-01-20.csv', 'data_Q1_2025/2025-01-21.csv', 'data_Q1_2025/2025-01-22.csv', 'data_Q1_2025/2025-01-23.csv', 'data_Q1_2025/2025-01-24.csv', 'data_Q1_2025/2025-01-25.csv', 'data_Q1_2025/2025-01-26.csv', 'data_Q1_2025/2025-01-27.csv', 'data_Q1_2025/2025-01-28.csv', 'data_Q1_2025/2025-01-29.csv', 'data_Q1_2025/2025-01-30.csv', 'data_Q1_2025/2025-01-31.csv', 'data_Q1_2025/2025-02-01.csv

In [39]:
zipfile.ZipFile("../data/zip/data_Q1_2025.zip", 'r').namelist()

['data_Q1_2025/2025-01-01.csv',
 'data_Q1_2025/2025-01-02.csv',
 'data_Q1_2025/2025-01-03.csv',
 'data_Q1_2025/2025-01-04.csv',
 'data_Q1_2025/2025-01-05.csv',
 'data_Q1_2025/2025-01-06.csv',
 'data_Q1_2025/2025-01-07.csv',
 'data_Q1_2025/2025-01-08.csv',
 'data_Q1_2025/2025-01-09.csv',
 'data_Q1_2025/2025-01-10.csv',
 'data_Q1_2025/2025-01-11.csv',
 'data_Q1_2025/2025-01-12.csv',
 'data_Q1_2025/2025-01-13.csv',
 'data_Q1_2025/2025-01-14.csv',
 'data_Q1_2025/2025-01-15.csv',
 'data_Q1_2025/2025-01-16.csv',
 'data_Q1_2025/2025-01-17.csv',
 'data_Q1_2025/2025-01-18.csv',
 'data_Q1_2025/2025-01-19.csv',
 'data_Q1_2025/2025-01-20.csv',
 'data_Q1_2025/2025-01-21.csv',
 'data_Q1_2025/2025-01-22.csv',
 'data_Q1_2025/2025-01-23.csv',
 'data_Q1_2025/2025-01-24.csv',
 'data_Q1_2025/2025-01-25.csv',
 'data_Q1_2025/2025-01-26.csv',
 'data_Q1_2025/2025-01-27.csv',
 'data_Q1_2025/2025-01-28.csv',
 'data_Q1_2025/2025-01-29.csv',
 'data_Q1_2025/2025-01-30.csv',
 'data_Q1_2025/2025-01-31.csv',
 'data_Q

In [33]:
def create_database_connection(
    pg_user="root", pg_pass="root", pg_host="localhost", pg_port=5432, pg_db="hard_drive_db"):
    """Create SQLAlchemy engine for PostgreSQL connection."""
    url = f'postgresql+psycopg://{pg_user}:{pg_pass}@{pg_host}:{pg_port}/{pg_db}'
    engine = create_engine(url)
    return engine


In [34]:
create_database_connection()

Engine(postgresql+psycopg://root:***@localhost:5432/hard_drive_db)